byzcoin/statetrie.go

```diff
@@ -0,0 +1,177 @@
+package byzcoin
```

Comment:
do you mean the snapshot feature? my understanding is that it'll be taken care of in a separate struct/function. the statetrie by itself doesn't store history.

Reply:
As discussed this morning: adding this functionality would require a big change to `trie`, so currently we'll just use the snapshot-feature. But I think it should go into `statetrie`, as the `trie` doesn't need to be aware of the snapshots, just give the possibility to do them...

Reply:
I see, in that case I think it is possible to extend `StateTrie` to support snapshots (just copy everything to another database/bucket) just by extending the API. You can have an additional field in the `StateTrie` struct that stores a pointer to the snapshot database. Then have a new function that executes the `CopyTo` function of `Trie` whenever a snapshot is needed. Perhaps we create a new bucket for every snapshot, maybe in a new database, the bucket name can be derived from skipblock ID so that we can quickly find what we're looking for or iterate through all the buckets to see what snapshots we have. The nested bucket feature - https://github.com/etcd-io/bbolt#nested-buckets might be useful here because we might store all the snapshots in one bucket but each snapshot in their own buckets.